### PR TITLE
fix: properly tear down partially initialized executors

### PR DESCRIPTION
**Pull request by FrederikBolding**
Refactor execution timeout logic to fix a problem where the execution worker would not be torn down correctly when failing to initialize. This PR also moves the entire initialization timeout logic to the `AbstractExecutionService`, simplifying the `SnapController`. Moving the timeout logic fixes problems where job initialization was not cancelled properly resulting in "Snap is already being executed", since execution of the async functions inside `AbstractExecutionService` would continue even after the timeout.

The ExecutionService gets an optional `maxInitTime` constructor argument that can be set to define init time for the specific service. This timeout value is spread between initiating the execution job and running the first execution of the Snap. If the execution environment fails to initiate we attempt to tear it down in case it has partially started (this is at least the case for iframes). We make sure that the timeout is applied to `initEnvStream` which has the smallest possible footprint and worry for side-effects. Additionally in case an error happens when terminating a Snap we catch that error an move on.

### CODE REVIEW

File: packages/snaps-controllers/src/services/AbstractExecutionService.ts [hunk 1]
```diff
@@ -55,6 +57,9 @@ export type Job<WorkerType> = {
   worker: WorkerType;
 };
 
+export type TerminateJobArgs<WorkerType> = Partial<Job<WorkerType>> &
```
Comment (hmalik88) [inline packages/snaps-controllers/src/services/AbstractExecutionService.ts:60]: why the change in type for the terminate job args?

Comment (FrederikBolding) [reply packages/snaps-controllers/src/services/AbstractExecutionService.ts:60]: I had to make the arguments partial to support termination of partially initialized executors, e.g. with the iframe executor the iframe has already been created in the DOM and we need to pass the ID. For most other executors this is not a problem.

https://github.com/MetaMask/snaps/pull/2348/files#diff-12564d49628a6405cf38ef5fdb16edd171449cb6f1f32301cdc068b94ae46cc2R252-R253

**Review by hmalik88 — APPROVED** [review]

**Review by Mrtenz — APPROVED** [review]
